File: Python/Class.py
#class juga memakai titik dua ':' ketimbang kurung kurawal '{}' seperti di C/C++/Java 
#body dari class juga harus di-indent
#alasan mengapa menggunakan self di setiap fungsinya, karena self ini menandakan objeknya, jadi ketika mengakses method tertentu secara implisit akan memasukkan reference objek tersebut ke method sebagai argumen, sehingga jika kita mengubah-ubah self tersebut, maka akan mengubah objek tersebut
#attributes secara default adalah public
from msvcrt import kbhit


class Class_ku:
    #class attributes dan bukan objek attributes. semacam static variable di dalam java
    # disini saya menyebutkan static variabel, namun sebenarnya di python disebutnya class variabel, mirip namun tidak identik, karena class variabel menggunakan namespace class
    class_var = 1
    #kita juga dapat menggunakan static typing, agar memastikan bahwa value yang dimasukan sebagai argumen adalah sesuai
    #dan dengan static typing ini, performa akan lebih cepat karena attributes objek akan lebih sedikit (semacam di filter terlebih dahulu pada saat memasukan argumennya)
    def __init__(self,nama,umur: int) -> None:
        #pendefinisian attributes objek. self kemudian nama attributs. jika kita mendifinisikan variabel tanpa self dan diluar __init__ maka akan dianggap sebagai class attributes bukan objek attributes
        self.nama=nama
        self.umur=umur

    def umur10TahunLalu(self):
        return self.umur-10

#end class
haris = Class_ku("haris",22)
print(haris.umur10TahunLalu())
print(Class_ku.class_var)


#mencoba inheritance
#inherit class dari python menggunakan kurung '()' ketimbang kata extends selayaknya di java atau C/C++
#semua method dan attributes parent class akan di inherit class anak
class Class_baru(Class_ku):
    def __init__(self, nama, umur: int,alamat:str) -> None:
        super().__init__(nama, umur)
        self.alamat = alamat
    
    def alamat_ku(self):
        return self.alamat

    #dapat menggunakan static variabel class parent atau class lain tanpa masalah
    #static method class
    def class_varku():
        return Class_ku.class_var
    def class_lain_var_ku():
        return Class_lain.class_lain_var

class Class_lain():
    class_lain_var = 1


haris_baru = Class_baru("haris_baru",22,"perumnas")
#akses method class_baru
print(haris_baru.alamat_ku())
#akses method class parent (CLass_ku)
print(haris_baru.umur10TahunLalu())
print(haris_baru.umur)
print(Class_baru.class_varku())

#error karena parameter pada method class_varku tidak ada self (objek yang memanggil), sehinngga tidak bisa dipanggil oleh objeknya
# print(haris_baru.class_varku())

print(Class_baru.class_lain_var_ku())

class Coba_static_variabel_lebih_lanjut:
    #setiap class yang memiliki attribute static, objek dari class tersebut akan mereference alamat attribute tersebut
    variabel_static = 69
    def __init__(self) -> None:
        pass

coba_static1 = Coba_static_variabel_lebih_lanjut()
coba_static2 = Coba_static_variabel_lebih_lanjut()

print(coba_static1.variabel_static)
print(coba_static2.variabel_static)

#jika nilai attribute static diubah dari class nya maka akan berubah pada objeknya pula, karena objek akan mereference alamat dari attribute class yang sama
Coba_static_variabel_lebih_lanjut.variabel_static=100000
print(coba_static1.variabel_static)
print(coba_static2.variabel_static)

#jika melakukan assignment ke variabel static, maka python akan membuat variabel dengan nama yang sama namun khusus untuk objek tersebut, sehingga tidak nilainya tidak mempengaruhi objek yang lain
coba_static1.variabel_static = 6969

#berbeda
print(coba_static1.variabel_static)
print(coba_static2.variabel_static)
print(Coba_static_variabel_lebih_lanjut.variabel_static)

class Mutable_static_variabel:
    #list termasuk mutable
    variabel_static = []
    def __init__(self) -> None:
        pass

# jika kita mengubah element dari variabel yang mutable oleh objek maka python tidak akan membuat variabel baru, sehingga variabel tersebut berubah untuk keseluruhan objek

coba_mutable_static_variabel1 = Mutable_static_variabel()
coba_mutable_static_variabel2 = Mutable_static_variabel()

coba_mutable_static_variabel1.variabel_static.append('a')

print(coba_mutable_static_variabel1.variabel_static)
print(coba_mutable_static_variabel2.variabel_static)


#private dan protected sebenarnya tidak ada dalam python, hanya saja jika pemrogram ingin mengatakan harus hati hati dalam mengurus attributes.
#private attributes hanya jika attribute tersebut diawali undersocre 2 kali (__) dan protected diawali underscore 1 kali (_)
#protected dapat diakses didalam package yang sama atau package yang berbeda namun subclass
#private tidak bisa diakses diluar class asal meskipun diakses oleh class anak, dipakai oleh objek pun tidak bisa. tetapi sebenarnya tetap bisa karena python akan melakukan "name mangling" yakni akan mengubah nama attributes private menjadi _(namaClass)(attributePrivate). attribute yang berubah ini dapat dilihat dengan __dict__
class Class_coba_private_protected:
    def __init__(self,_nama,_umur:int,__jenis_kelamin:bool) -> None:
        self._nama = _nama
        self._umur = _umur
        self.__jenis_kelamin = __jenis_kelamin

    def get_jenis_kelamin(self):
        return self.__jenis_kelamin

objek_coba = Class_coba_private_protected("Haris",21,True)
#menggunakan ternary operator
#protected bisa diakses pada objek di package yang sama 
print("Jenis kelamin "+objek_coba._nama+" adalah laki" if objek_coba.get_jenis_kelamin() else "Jenis kelamin "+objek_coba._nama+" adalah perempuan","dan berumur "+str(objek_coba._umur)+" tahun")

#error
# print(objek_coba.__jenis_kelamin)

#tidak error karena mengakses private attributes melalui nama yang sudah dimangling
print(objek_coba._Class_coba_private_protected__jenis_kelamin)

#melihat name mangling
print(objek_coba.__dict__)

class Anak_class_coba_private_protected(Class_coba_private_protected):
    def __init__(self, _nama, _umur: int, __jenis_kelamin: bool,_attribute_baru) -> None:
        super().__init__(_nama, _umur, __jenis_kelamin)
        self._attribute_baru = _attribute_baru

File: JavaScript/Class.js
//class adalah kasus spesial dari function
//class tidak boleh memiliki lebih dari 1 constructor
//class adalah spesial function, sehingga karena function memiliki prototype Function.prototype->Object.prototype->null
class haris{
    //constructor juga bisa memiliki parameter default
    constructor(x,y=1){
        this.x = x;
        this.y = y;
    }
    //method haris
    methodku(){
        console.log(this.x+this.y);
    }
}

var objectHaris = new haris(1,2);
var objectHarisDefaultY = new haris(4);

// Sebenarnya kita dapat mendefinisikan juga objek dengan funciton biasa
function harisPakeFunction(x,y){
    this.x=x;
    this.y=y;
    //bisa menggunakan arrow function 
    this.methodku = ()=>console.log("this.x+this.y = "+(this.x+this.y));
}

let objectHarisPakeFunction = new harisPakeFunction(1,2);
objectHarisPakeFunction.methodku();


//class dapat memiliki field dan method
//dalam class haris ada 1 method yaitu methodku sedangkan untuk field nya adalah x dan y
console.log(objectHaris.y);
objectHaris.methodku();
console.log(objectHarisDefaultY.y);


//class dapat memiliki static method, field maupun static initializier block, sehingga static method maupun field tidak bisa dipanggil
//melalui instance class tersebut
//sebuah method static atau field static suatu class tidak dimiliki oleh objek yang merupakan objek class tersebut
//tapi static ini milik class bukan objek
class coba{
    //x akan dieksekusi terlebih dahulu
    static x=11;
    //static initializer block
    static {
        console.log("x dalam static initializer block adalah"+coba.x);
    }
    static haha(){
        console.log("ini x di dalam static haha "+coba.x);
    }

}

//static field dan static initializer block akan dijalankan sesuai urutan dalam penulisan dalam class

//untuk memanggil method static dalam class adalah (class).(methodstatic) seperti java
coba.haha();

//dalam class ada class declaration seperti diatas dan class expression, sama seperti function
//hanya saja untuk masalah hoisting pada class declaration, class dapat digunakan semenjak class didefinisikan, jadi tidak bisa sembarang
//tidak seperti function declaration bisa ditempatkan di sembarang tempat

//error
/*var hoisting = new hoist();
class hoist{
    hehe = 0;
}*/

let classExpression = class {
    coba = 0;
    constructor(coba=10) {
        this.coba = coba;
    }
}

//class expression berperilaku sebagaimana function expression
//undefined
console.log(classExpression.coba)

let objectDariClassExpression = new classExpression();
console.log(objectDariClassExpression.coba)

//untuk menjadikan suatu field adalah private, harus ditambah # di depan fieldnya
//class dalam javascript juga bisa memiliki getter dan setter, tetapi agak unik
//karena biasanya dalam java adalah bernama get(nama fieldnya), namun dalam javascript bisa nama apa saja tetapi harus diawali
//dengan get. begitu juga untuk set
//dan yang lebih unik lagi nama get dan set bisa sama seperti dibawah
//untuk getnya bisa tanpa parentheses contohnya objek.x
//dan untuk settnya juga menggunakan assignment contohnya objek.x=10 maka 10 akan dimasukkan ke cobalagi
let classPrivate = class {
    //untuk menandakan bahwa #cobalagi adalah number
    #cobalagi=0;
    constructor(cobalagi){
        this.#cobalagi=cobalagi;
    }
    get x(){
        return this.#cobalagi;
    }
    set x(cobalagi){
        this.#cobalagi = cobalagi;
    }
}

let objectDariClassPrivate = new classPrivate(69);

//error karena #cobalagi adalah private
/*console.log(objectDariClassPrivate.#cobalagi)*/

//memasukkan nilai 10101010 ke cobalagi
objectDariClassPrivate.x=10101010;
console.log(objectDariClassPrivate.x)

//prototype dari class adalah Function
console.log(classPrivate.prototype.prototype);

//objek yang juga dapat didefinisikan suatu method atau attribute baru dengan memasukkannya ke prototype
//jika method atau attribute dimasukkan ke prototype, maka objek lain dari class yang sama tidak ada
function classHehe(x=1,y=2){
    this.x=x;
    this.y=y;
    
};

classHehe.prototype.methodBaru= ()=>{console.log("hahahaha")};

let objekDenganMethodBaruDariPrototype = new classHehe();
objekDenganMethodBaruDariPrototype.methodBaru();

//class dalam javascript bisa private juga. Private bisa dengan menambahkan '_' atau '#' sebagai prefix attribute
class cobaPrivate{
    #private2=6;
    _mmm=2;
    constructor(){
        //menggunakan private identifier '_' dengan this tidak akan error
        this._private = 3;
        //menggunakan this dengan private identifier '#' akan error
        // this.#kkk=2
    }
    method(){
        console.log(this._private);
        console.log(this.#private2);
    }
}

let cobaPrivateObjek = new cobaPrivate();
cobaPrivateObjek.method();

//class dalam javascript juga bisa untuk inheritance
class parentClass{
    constructor(x=10){
        this.hehe=x;
    }
    aaa(){
        console.log(this.hehe);
    }
}

class childClass extends parentClass{
    constructor(y=10){
        super(y);
        this._attributeChild=90;
    }
}

let objekChildClass = new childClass(90);

//child objek bisa mengakses method dari parent class.
objekChildClass.aaa();